### ISSUE TITLE
Some routes don't work, others do, no reason to be found

### ISSUE THREAD

**Issue by ChrGriffin** [nWidart/laravel-modules#241]
I created three modules:

`php artisan module:make Finances Jobs Etsy`

And the routing only seems to be working for the 'Jobs' module. Finances root and Etsy root both return a NotFoundHttpException, however the Jobs module root will return the default generated view.

I added the auth middleware and an additional prefix to each one's route configuration:

`Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/finances', 'namespace' => 'Modules\Finances\Http\Controllers']...`
`Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/jobs', 'namespace' => 'Modules\Jobs\Http\Controllers']...`
`Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/etsy', 'namespace' => 'Modules\Etsy\Http\Controllers']...`

As you can see they are identical aside from the module name itself.

I tried:

`php artisan route:clear`
`php artisan cache:clear`

however neither one has solved the issue.

Running:

`php artisan route:list`

Shows that the routes DO exist, and are being found by Laravel. And running:

` php artisan module:list`

DOES show all three modules as existing.

What could be causing these routes to fail for two modules, but not a third, where the routes are treated identically in each case?

**Comment by nWidart** [comment]
Have you check the routes file of the 3 modules ? Is the routes file of the first 2 module not generated properly ?

**Comment by ChrGriffin** [comment]
Here's the route file for the Jobs module (the one that's working):

`<?php

Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/jobs', 'namespace' => 'Modules\Jobs\Http\Controllers'], function()
{
    Route::get('/', 'JobsController@index');
});`

And here are the route files for the other two:

`<?php

Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/finances', 'namespace' => 'Modules\Finances\Http\Controllers'], function()
{
    Route::get('/', 'FinancesController@index');
});`
`<?php

Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/etsy', 'namespace' => 'Modules\Etsy\Http\Controllers'], function()
{
    Route::get('/', 'EtsyController@index');
});`

They seem identical to me (again, aside from the module name of course). In any case the routes DO show when one runs:

`php artisan route:list`

so I'd assume that means they're generated correctly?

**Comment by nWidart** [comment]
They indeed all look fine.

You can try to change the route names in place of the `/`. Just general debugging to understand why 2 aren't caught.

**Comment by ChrGriffin** [comment]
I tried changing the `/` root path to `/dashboard` for both, cleared cache, then tried. No luck unfortunately.

**Comment by nWidart** [comment]
did you try to `dd()` inside the route files?

**Comment by ChrGriffin** [comment]
I did, however I can't actually get anywhere that the route would be dumped, since trying to navigate to, say, the Finances model, just gives me a `NotFoundHttpException in RouteCollection.php line 161:` error.

**Comment by nWidart** [comment]
You must have an error somewhere in your route definition and what you're trying to access in the browser. Just follow the laravel documentation to handle your routes.

I'll close this for now as it's general laravel related issue.

**Comment by ChrGriffin** [comment]
I must disagree. If there's an error with the route definition then it's an error that was generated by the `php artisan module:make` command, as the _only_ change I've made to the routes has been to add the auth middleware and a slight change to the generated prefixes, and in any case they didn't work before then either.

However, I seriously doubt it's an issue with that. The base app route file is extremely simple:

```
// home page
Route::get('/', function () {
    return view('welcome');
});

// login, logout, register, etc.
Route::auth();

// admin dashboard
Route::get('/admin', 'AdminController@index');
```

As is the Finances module routes file:

```
Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/finances', 'namespace' => 'Modules\Finances\Http\Controllers'], function()
{
    Route::get('/', 'FinancesController@index');
});
```

And the URL:

```
https://[site url]/admin/finances
```

I'm afraid I don't see where or how, in this very simple setup, the error is on my end, _or_ with the routes themselves. Particularly when another module with the exact same setup seems to work perfectly. _Especially_, when the routes _do_ appear listed when using `php artisan route:list`. Every indication, besides the page not loading, is that everything is set up correctly.

**Comment by nWidart** [comment]
It's because this package does _nothing_ to your routes. You have full control over how they're loaded etc. You can debug this by the previously mentioned ways.

**Comment by ChrGriffin** [comment]
I've done some more experimenting and finally nailed down a pattern to this issue. The issue is _not_ with the routes, but with the autoloading - though how, I can't begin to imagine. One remembers how the 'Jobs' module worked, but the other two did not? Well, to test, I added a fourth planned module, 'Recipes', which worked! ...and caused 'Jobs' to stop working. Once I deleted both 'Jobs' and 'Recipes', 'Finances' began to work.

In short, it seems that the module that is _last_, alphabetically, is the only one being autoloaded. Removing the last module alphabetically causes the next to work, and the inverse. Again, for reasons I can't even begin to fathom.

**Comment by nWidart** [comment]
Try using the route service provider to load your routes as I suggested before.

**Comment by ChrGriffin** [comment]
I did say yes, because they _do_ appear when you run `php artisan route:list`. It seems the issue is the web application specifically not autoloading the modules.

How would I load my routes in the service provider?

**Comment by nWidart** [comment]
Check the module:make route service provider command. That's only an example though, you can load your routes in any service provider you like in whatever way you prefer.

**Comment by ChrGriffin** [comment]
I gave it a shot. I made a service provider for the jobs module, and in the register function I tried both:

```
require(dirname(__FILE__) . '/../Http/routes.php');
```

and

```
require(dirname(__FILE__) . '/../start.php');
```

Behaviour was unchanged in both cases.

**Comment by ChrGriffin** [comment]
A further update: after a _lot_ of playing around with service providers, routes, debugging, etc. I've finally determined the root issue appears to be this route, in my main app routes file:

```
Route::get('/admin', 'AdminController@index');
```

I can only assume, because all the other routes had the prefix 'admin' on them:

```
Route::group(['middleware' => ['web', 'auth'], 'prefix' => 'admin/finances', 'namespace' => 'Modules\Finances\Http\Controllers'], function()
{
    Route::get('/', 'FinancesController@index');
});
```

Why this resulted in the very specific and odd behaviour that was encountered here - only the last module, alphabetically, having working routes, and the rest not - I'm not remotely sure.

**Comment by ChrGriffin** [comment]
Not a clue ¯\_(ツ)_/¯ I needed a new environment anyway, so ultimately I 'fixed' it by rebuilding my application 5.6 (it was in 5.4) and deploying it on a new server.
